File: java/problem238.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];

        int p = 1;
        for (int i=0;i<n;i++){
            ans[i] = p;
            p *= nums[i];
        }

        int s = 1;
        for (int i=n-1;i >= 0;i--){
            ans[i] *= s;
            s *= nums[i];
        }
        return ans;
    }
}

File: python/problem238.py
class Solution(object):
    def productExceptSelf(self, nums):
        n = len(nums)
        ans = [1] * n

        prefix = 1
        for i in range(n):
            ans[i] = prefix
            prefix *= nums[i]
            print(prefix)

        suffix = 1
        for i in range(n - 1, -1, -1):
            ans[i] *= suffix
            suffix *= nums[i]
            print(suffix)

        return ans